overlord/servicestate/quota_control.go

```diff
@@ -266,6 +292,38 @@ func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) (*st
 	return ts, nil
 }
 
+// remove a string item at index i from the string slice,
+// it maintains the ordering of the original slice.
+func remove(slice []string, i int) []string {
+	return append(slice[:i], slice[i+1:]...)
+}
+
+// ensureSnapServicesAbsentFromSubGroups removes all service references of a snap in
+// sub-groups related to the group of the snap, and returns the groups that were modified.
+func ensureSnapServicesAbsentFromSubGroups(grp *quota.Group, snap string, allGrps map[string]*quota.Group) ([]*quota.Group, error) {
+	// We can assume that when removing a snap, that if it has services in
+	// sub-groups that there will be only one level of sub-groups, so we can
+	// avoid checking further sub-groups.
+	var modifiedGrps []*quota.Group
+	for _, name := range grp.SubGroups {
+		subgrp, ok := allGrps[name]
+		if !ok {
+			return nil, fmt.Errorf("non-existent sub-group %q", name)
+		}
+
+		for idx, svc := range subgrp.Services {
+			parts := strings.Split(svc, ".")
+			if parts[0] == snap {
+				// found a service that matches the snap we are removing,
+				// so remove that too
+				subgrp.Services = remove(subgrp.Services, idx)
```

Comment:
Can it really be that in a group you have some services from a snap and some from another? If not, then once you find a matching service, you can wipe out the while `subgrp.Services` list.

Reply:
It's legal to do so right now, I totally see your point about this. I don't know if there will be a use-case for this behavior, but at-least according to the spec nothing prohibits you for doing this right now.